File: hw2-apriori-algorithm/main.cpp
#include <algorithm>
#include <cstdio>
#include <cstring>
#include <ctime>
#include <iostream>
#include <set>
#include <unordered_map>
#include <vector>
using namespace std;

struct Node {
	unsigned int val, level;
	unordered_map<int, Node *> child;
};

char inputpath[100];
char outputpath[100];
FILE *fin, *fout;
unordered_map<int, int> C1;
vector<int> tempv, Lsup;
unordered_map<Node *, int> Csup;
vector<vector<int>> Cits, Ctemp, Lits;
set<vector<int>> Lset;
int tempa, tempb;
unsigned int tempn, cnt;
Node *root;

unsigned int grouplen = 1, support;

void generateC1();
void generateL1();
void outputFile();
void generateCtemp();
void generateC();
void generateCsup();
void generateL();

int main(int argc, char *argv[]) {
	if (argc < 4) {
		printf("Missing parameter\n");
		string temps;
		printf("file input:");
		getline(cin, temps);
		strcpy(inputpath, temps.c_str());
		printf("file output: ");
		getline(cin, temps);
		strcpy(outputpath, temps.c_str());
		printf("support: ");
		cin >> support;
	} else {
		strcpy(inputpath, argv[1]);
		strcpy(outputpath, argv[2]);
		support = atoi(argv[3]);
	}
	printf("file input: %s\n", inputpath);
	printf("file output: %s\n", outputpath);
	printf("support: %d\n", support);

	clock_t start_time = clock();

	fin = fopen(inputpath, "rb");
	if (fin == NULL) {
		cerr << "Fail to open input file";
		return 1;
	}
	fout = fopen(outputpath, "w");
	if (fout == NULL) {
		cerr << "Fail to open output file";
		return 1;
	}

	generateC1();
	generateL1();

	int Llensum = 0;
	while (Lits.size()) {
		outputFile();

		printf("L%d: %d\n", grouplen, Lits.size());
		Llensum += Lits.size();

		generateCtemp();
		printf("C%dtemp: %d\n", grouplen + 1, Ctemp.size());

		grouplen++;

		generateC();
		printf("C%dits: %d\n", grouplen, Cits.size());

		generateCsup();

		generateL();
	}

	fclose(fout);

	printf("Lall: %d\n", Llensum);
	printf("It tooks %d milliseconds\n", (clock() - start_time) * 1000 / CLOCKS_PER_SEC);
}

inline unsigned int readint(FILE *&file) {
	fread(&tempa, 1, 4, file);
	return tempa;
}

void dfs(Node *&now, vector<int> item) {
	if (now->level == grouplen) {
		if (Csup[now] >= support) {
			Lits.push_back(item);
			Lsup.push_back(Csup[now]);
		}
	} else {
		for (auto &next : now->child) {
			vector<int> nextitem = item;
			nextitem.push_back(next.first);
			dfs(next.second, nextitem);
		}
	}
}

void generateC1() {
	while (true) {
		tempn = readint(fin);
		if (feof(fin)) break;
		readint(fin);
		cnt = readint(fin);
		while (cnt--) {
			tempn = readint(fin);
			C1[tempn]++;
		}
	}
	fclose(fin);
}

void generateL1() {
	Lsup.clear();
	for (auto &v : C1) {
		if (v.second >= support) {
			Lits.push_back(vector<int>(1, v.first));
			Lsup.push_back(v.second);
		}
	}
}

void outputFile() {
	for (int i = 0; i < Lits.size(); i++) {
		fprintf(fout, "%d", Lits[i][0]);
		for (int j = 1; j < grouplen; j++) {
			fprintf(fout, ",%d", Lits[i][j]);
		}
		fprintf(fout, ":%d\n", Lsup[i]);
	}
}

void generateCtemp() {
	Ctemp.clear();

	// cross join
	for (int i = 0; i < Lits.size(); i++) {
		for (int j = i + 1; j < Lits.size(); j++) {
			int k = 0;
			while (k < grouplen - 1 && Lits[i][k] == Lits[j][k]) k++;
			if (k == grouplen - 1) {
				Ctemp.push_back(Lits[i]);
				tempa = Lits[i][grouplen - 1];
				tempb = Lits[j][grouplen - 1];
				if (tempa > tempb) swap(tempa, tempb);
				Ctemp[Ctemp.size() - 1][grouplen - 1] = tempa;
				Ctemp[Ctemp.size() - 1].push_back(tempb);
			}
		}
	}
}

void generateC() {
	Lset.clear();
	for (auto &itemset : Lits) {
		Lset.insert(itemset);
	}
	Cits.clear();
	for (auto &itemset : Ctemp) {
		int i = 0;
		while (i < grouplen) {
			tempv = itemset;
			tempv.erase(tempv.begin() + i);
			if (Lset.find(tempv) == Lset.end()) {  // https://en.cppreference.com/w/cpp/container/set/find
				break;
			}
			i++;
		}
		if (i == grouplen) {
			Cits.push_back(itemset);
		}
	}
}

void generateCsup() {
	Csup.clear();

	Node *now;
	root = new Node();
	root->val = -1;
	root->level = 0;
	for (auto &item : Cits) {
		now = root;
		for (int j = 0; j < grouplen; j++) {
			if (now->child.find(item[j]) == now->child.end()) {
				now->child[item[j]] = new Node();
				now->child[item[j]]->val = item[j];
				now->child[item[j]]->level = j + 1;
			}
			now = now->child[item[j]];
		}
		Csup[now] = 0;
	}

	// cnt sup
	fin = fopen(inputpath, "rb");
	if (fin == NULL) {
		fprintf(stderr, "Fail to open input file");
		exit(1);
	}
	unsigned int cnt;
	vector<Node *> nownode;
	while (true) {
		tempn = readint(fin);
		if (feof(fin)) break;
		readint(fin);
		cnt = readint(fin);
		nownode.clear();
		nownode.push_back(root);
		while (cnt--) {
			tempn = readint(fin);
			for (int i = nownode.size() - 1; i >= 0; i--) {
				if (nownode[i]->child.find(tempn) != nownode[i]->child.end()) {
					if (nownode[i]->child[tempn]->level == grouplen) {
						Csup[nownode[i]->child[tempn]]++;
					} else {
						nownode.push_back(nownode[i]->child[tempn]);
					}
				}
			}
		}
	}
	fclose(fin);
}

void generateL() {
	// filter support
	Lits.clear();
	Lsup.clear();
	dfs(root, vector<int>());
}

File: hw2-apriori-algorithm/main.py
import argparse
import time


parser = argparse.ArgumentParser()
parser.add_argument('input', default=None)
parser.add_argument('output', default=None)
parser.add_argument('support', default=None, type=int)
args = parser.parse_args()

start_time = time.time()

inputpath = args.input
outputpath = args.output
support = args.support
print('file input: {}'.format(inputpath))
print('file output: {}'.format(outputpath))
print('support: {}'.format(support))

fin = open(inputpath, 'rb')
fout = open(outputpath, 'w')


def dfs(now, item):
    if now['level'] == grouplen:
        if now['support'] >= support:
            Lits.append(item.copy())
            Lsup.append(now['support'])
    else:
        for nextitemid in now['child']:
            nextitem = item.copy()
            nextitem.append(nextitemid)
            dfs(now['child'][nextitemid], nextitem)


def readint():
    buf = fin.read(4)
    if not buf:
        return None
    return int.from_bytes(buf, byteorder='little')


Cits = []
Csup = {}
grouplen = 1
C1 = {}

while True:
    tempn = readint()
    if tempn is None:
        break
    readint()
    cnt = readint()
    while cnt > 0:
        tempn = readint()
        if tempn not in C1:
            C1[tempn] = 0
        C1[tempn] += 1
        cnt -= 1
fin.close()

Lits = []
Lsup = []
for item in C1:
    if C1[item] >= support:
        Lits.append([item])
        Lsup.append(C1[item])

Llensum = 0
while len(Lits) > 0:
    # output file
    for i in range(len(Lits)):
        fout.write('{}'.format(Lits[i][0]))
        for j in range(1, grouplen):
            fout.write(',{}'.format(Lits[i][j]))
        fout.write(':{}\n'.format(Lsup[i]))
    print('L{}: {}'.format(grouplen, len(Lits)))
    Llensum += len(Lits)

    Ctemp = []
    Ctemplen = 0
    # cross join
    for i in range(len(Lits)):
        for j in range(i + 1, len(Lits)):
            k = 0
            while k < grouplen - 1 and Lits[i][k] == Lits[j][k]:
                k += 1
            if k == grouplen - 1:
                Ctemp.append(Lits[i].copy())
                Ctemplen += 1
                tempa = Lits[i][grouplen - 1]
                tempb = Lits[j][grouplen - 1]
                if tempa > tempb:
                    tempa, tempb = tempb, tempa
                Ctemp[Ctemplen - 1][grouplen - 1] = tempa
                Ctemp[Ctemplen - 1].append(tempb)

    print('C{}temp: {}'.format(grouplen + 1, len(Ctemp)))

    grouplen += 1

    # remove not in Lits
    Lset = set()
    for i in range(len(Lits)):
        Lset.add(tuple(Lits[i]))
    Cits.clear()
    for i in range(len(Ctemp)):
        j = 0
        while j < grouplen:
            if tuple(Ctemp[i][:j] + Ctemp[i][j + 1:]) not in Lset:
                break
            j += 1
        if j == grouplen:
            Cits.append(Ctemp[i])

    print('C{}dits: {}'.format(grouplen, len(Cits)))

    Csup.clear()

    root = {'val': -1, 'level': 0, 'child': {}}
    for i in range(len(Cits)):
        now = root
        for j in range(grouplen):
            if Cits[i][j] not in now['child']:
                now['child'][Cits[i][j]] = {
                    'val': Cits[i][j],
                    'level': j + 1,
                    'child': {}
                }
            now = now['child'][Cits[i][j]]
        now['support'] = 0

    # cnt sup
    fin = open(inputpath, 'rb')
    nownode = []
    while True:
        tempn = readint()
        if tempn is None:
            break
        readint()
        cnt = readint()
        nownode.clear()
        nownode.append(root)
        while cnt > 0:
            tempn = readint()
            for i in range(len(nownode) - 1, -1, -1):
                if tempn in nownode[i]['child']:
                    if nownode[i]['child'][tempn]['level'] == grouplen:
                        nownode[i]['child'][tempn]['support'] += 1
                    else:
                        nownode.append(nownode[i]['child'][tempn])
            cnt -= 1
    fin.close()

    Lits.clear()
    Lsup.clear()
    dfs(root, [])

fout.close()

print('Lall: {}'.format(Llensum))
print('It tooks {0:.0f} milliseconds'.format(
    (time.time() - start_time) * 1000))
